File: backend/api/consumption.py
"""Consumption API endpoints."""
import csv
import io
from flask import Blueprint, request, jsonify, Response
from datetime import datetime

from backend.services.consumption_service import (
    get_consumption,
    aggregate_by_granularity,
    filter_consumption,
    aggregate_by_dimension,
    calculate_totals
)
from backend.utils.date_validators import validate_date_range
from backend.config.settings import SUPPORTED_REGIONS
from backend.utils.errors import APIError
from backend.middleware.auth_middleware import require_auth
from backend.database import SessionLocal
from backend.models.user import User


consumption_bp = Blueprint('consumption', __name__)


from backend.utils.session_helpers import get_account_id_from_session


@consumption_bp.route('/consumption', methods=['GET'])
@require_auth
def get_consumption_endpoint():
    """
    Get consumption data for a date range.
    Authentication required.
    
    Query parameters:
        - from_date: Start date (required, ISO format: YYYY-MM-DD)
        - to_date: End date (required, ISO format: YYYY-MM-DD)
        - granularity: "day", "week", "month" (optional, default: "day")
        - region: Filter by region (optional)
        - service: Filter by service name (optional)
        - resource_type: Filter by resource type (optional)
        - aggregate_by: "resource_type", "region", "tag" (optional)
        - force_refresh: Force refresh cache (optional: true/false)
    """
    session = getattr(request, 'session', None)
    if not session:
        raise APIError("Authentication required", status_code=401)
    
    # Get query parameters
    from_date = request.args.get('from_date')
    to_date = request.args.get('to_date')
    granularity = request.args.get('granularity', 'day')
    region = request.args.get('region')
    service = request.args.get('service')
    resource_type = request.args.get('resource_type')
    aggregate_by = request.args.get('aggregate_by')
    force_refresh = request.args.get('force_refresh', 'false').lower() == 'true'
    
    # Validate required parameters
    if not from_date or not to_date:
        raise APIError("from_date and to_date parameters are required", status_code=400)
    
    # Validate date range using centralized validator
    is_valid, error_msg = validate_date_range(from_date, to_date, granularity)
    if not is_valid:
        raise APIError(error_msg, status_code=400)
    
    # Validate granularity
    if granularity not in ['day', 'week', 'month']:
        raise APIError("granularity must be 'day', 'week', or 'month'", status_code=400)
    
    # Validate region if provided
    if region and region not in SUPPORTED_REGIONS:
        raise APIError(
            f"Unsupported region: {region}. Supported regions: {', '.join(SUPPORTED_REGIONS)}",
            status_code=400
        )
    
    # Get account_id from session
    account_id = get_account_id_from_session()
    if not account_id:
        raise APIError("Could not retrieve account information", status_code=401)
    
    # Use session region if region not specified
    query_region = region or session.region
    
    try:
        # Get consumption data
        consumption_data = get_consumption(
            access_key=session.access_key,
            secret_key=session.secret_key,
            region=query_region,
            account_id=account_id,
            from_date=from_date,
            to_date=to_date,
            force_refresh=force_refresh
        )
        
        # Apply filters
        if region or service or resource_type:
            consumption_data = filter_consumption(
                consumption_data,
                region=region,
                service=service,
                resource_type=resource_type
            )
        
        # Aggregate by granularity if specified
        if granularity != 'day' or aggregate_by:
            if aggregate_by:
                # Aggregate by dimension
                aggregated = aggregate_by_dimension(consumption_data, aggregate_by)
                consumption_data = {
                    **consumption_data,
                    "entries": aggregated,
                    "aggregated_by": aggregate_by
                }
            else:
                # Aggregate by granularity
                aggregated = aggregate_by_granularity(consumption_data, granularity)
                consumption_data = {
                    **consumption_data,
                    "entries": aggregated,
                    "granularity": granularity
                }
        
        # Calculate totals
        totals = calculate_totals(consumption_data)
        
        # Get currency from consumption data if available
        currency = consumption_data.get("currency")
        
        return jsonify({
            "success": True,
            "data": consumption_data,
            "totals": totals,
            "metadata": {
                "from_date": from_date,
                "to_date": to_date,
                "region": query_region,
                "currency": currency,
                "granularity": granularity,
                "filters": {
                    "region": region,
                    "service": service,
                    "resource_type": resource_type
                },
                "aggregate_by": aggregate_by
            }
        }), 200
    
    except ValueError as e:
        raise APIError(str(e), status_code=400)
    except Exception as e:
        raise APIError(f"Failed to fetch consumption: {str(e)}", status_code=500)


@consumption_bp.route('/consumption/export', methods=['GET'])
@require_auth
def export_consumption_endpoint():
    """
    Export consumption data to CSV or JSON.
    Authentication required.
    
    Query parameters:
        - Same as /consumption endpoint
        - format: "csv" or "json" (required)
    """
    session = getattr(request, 'session', None)
    if not session:
        raise APIError("Authentication required", status_code=401)
    
    export_format = request.args.get('format', 'json').lower()
    
    if export_format not in ['csv', 'json']:
        raise APIError("format must be 'csv' or 'json'", status_code=400)
    
    # Get consumption data using same logic as main endpoint
    from_date = request.args.get('from_date')
    to_date = request.args.get('to_date')
    granularity = request.args.get('granularity', 'day')
    region = request.args.get('region')
    service = request.args.get('service')
    resource_type = request.args.get('resource_type')
    aggregate_by = request.args.get('aggregate_by')
    force_refresh = request.args.get('force_refresh', 'false').lower() == 'true'
    
    if not from_date or not to_date:
        raise APIError("from_date and to_date parameters are required", status_code=400)
    
    # Get account_id from session
    account_id = get_account_id_from_session()
    if not account_id:
        raise APIError("Could not retrieve account information", status_code=401)
    
    query_region = region or session.region
    
    try:
        # Get consumption data
        consumption_data = get_consumption(
            access_key=session.access_key,
            secret_key=session.secret_key,
            region=query_region,
            account_id=account_id,
            from_date=from_date,
            to_date=to_date,
            force_refresh=force_refresh
        )
        
        # Apply filters
        if region or service or resource_type:
            consumption_data = filter_consumption(
                consumption_data,
                region=region,
                service=service,
                resource_type=resource_type
            )
        
        # Aggregate if needed
        if aggregate_by:
            aggregated = aggregate_by_dimension(consumption_data, aggregate_by)
            consumption_data = {
                **consumption_data,
                "entries": aggregated
            }
        elif granularity != 'day':
            aggregated = aggregate_by_granularity(consumption_data, granularity)
            consumption_data = {
                **consumption_data,
                "entries": aggregated
            }
        
        entries = consumption_data.get("entries", [])
        
        if export_format == 'csv':
            # Generate CSV
            output = io.StringIO()
            if entries:
                # Get all unique keys from entries
                fieldnames = set()
                for entry in entries:
                    fieldnames.update(entry.keys())
                fieldnames = sorted(fieldnames)
                
                writer = csv.DictWriter(output, fieldnames=fieldnames)
                writer.writeheader()
                writer.writerows(entries)
            
            return Response(
                output.getvalue(),
                mimetype='text/csv',
                headers={
                    'Content-Disposition': f'attachment; filename=consumption_{from_date}_to_{to_date}.csv'
                }
            )
        else:
            # Return JSON
            return jsonify({
                "success": True,
                "data": consumption_data
            }), 200
    
    except Exception as e:
        raise APIError(f"Failed to export consumption: {str(e)}", status_code=500)

File: frontend/js/consumption.js
/**
 * Consumption service module for OSC-FinOps
 * Handles API communication for consumption history
 */

const ConsumptionService = {
    API_BASE: '/api',
    currentData: null,

    /**
     * Get session ID from localStorage
     */
    getSessionId() {
        return localStorage.getItem('osc_finops_session_id');
    },

    /**
     * Build headers with session ID
     */
    getHeaders() {
        const headers = {
            'Content-Type': 'application/json'
        };
        const sessionId = this.getSessionId();
        if (sessionId) {
            headers['X-Session-ID'] = sessionId;
        }
        return headers;
    },

    /**
     * Get consumption data
     * 
     * Note: ReadAccountConsumption returns pre-aggregated data:
     * - Data separated by type (each resource/service type has its own entry)
     * - Consolidated quantity over the queried period
     * - Unit price (does not vary with period)
     * - Total cost per type already calculated (quantity × unit_price) in Price field
     * 
     * @param {Object} params - Query parameters
     * @param {string} params.from_date - Start date (YYYY-MM-DD) - inclusive
     * @param {string} params.to_date - End date (YYYY-MM-DD) - exclusive
     * @param {string} [params.granularity] - Granularity: 'day', 'week', 'month'
     * @param {string} [params.region] - Filter by region
     * @param {string} [params.service] - Filter by service
     * @param {string} [params.resource_type] - Filter by resource type
     * @param {string} [params.aggregate_by] - Aggregate by: 'resource_type', 'region', 'tag'
     * @param {boolean} [params.force_refresh] - Force cache refresh
     * @returns {Promise<Object>} Consumption data with pre-aggregated entries
     */
    async getConsumption(params) {
        try {
            const queryParams = new URLSearchParams();
            
            // Required parameters
            if (params.from_date) queryParams.append('from_date', params.from_date);
            if (params.to_date) queryParams.append('to_date', params.to_date);
            
            // Optional parameters
            if (params.granularity) queryParams.append('granularity', params.granularity);
            if (params.region) queryParams.append('region', params.region);
            if (params.service) queryParams.append('service', params.service);
            if (params.resource_type) queryParams.append('resource_type', params.resource_type);
            if (params.aggregate_by) queryParams.append('aggregate_by', params.aggregate_by);
            if (params.force_refresh) queryParams.append('force_refresh', 'true');
            
            const response = await fetch(`${this.API_BASE}/consumption?${queryParams.toString()}`, {
                method: 'GET',
                headers: this.getHeaders()
            });
            
            const data = await response.json();
            
            if (!response.ok) {
                throw new Error(data.error?.message || `HTTP ${response.status}: ${response.statusText}`);
            }
            
            if (data.success) {
                this.currentData = data.data;
                return data;
            } else {
                throw new Error(data.error?.message || 'Failed to fetch consumption data');
            }
        } catch (error) {
            console.error('Get consumption error:', error);
            throw error;
        }
    },

    /**
     * Export consumption data
     * @param {Object} params - Same as getConsumption
     * @param {string} format - Export format: 'csv' or 'json'
     * @returns {Promise<Blob|Object>} Exported data
     */
    async exportConsumption(params, format = 'json') {
        try {
            const queryParams = new URLSearchParams();
            
            // Required parameters
            if (params.from_date) queryParams.append('from_date', params.from_date);
            if (params.to_date) queryParams.append('to_date', params.to_date);
            
            // Optional parameters
            if (params.granularity) queryParams.append('granularity', params.granularity);
            if (params.region) queryParams.append('region', params.region);
            if (params.service) queryParams.append('service', params.service);
            if (params.resource_type) queryParams.append('resource_type', params.resource_type);
            if (params.aggregate_by) queryParams.append('aggregate_by', params.aggregate_by);
            if (params.force_refresh) queryParams.append('force_refresh', 'true');
            
            queryParams.append('format', format);
            
            const response = await fetch(`${this.API_BASE}/consumption/export?${queryParams.toString()}`, {
                method: 'GET',
                headers: this.getHeaders()
            });
            
            if (!response.ok) {
                const errorData = await response.json();
                throw new Error(errorData.error?.message || `HTTP ${response.status}: ${response.statusText}`);
            }
            
            if (format === 'csv') {
                // Return blob for CSV download
                const blob = await response.blob();
                return blob;
            } else {
                // Return JSON
                const data = await response.json();
                return data;
            }
        } catch (error) {
            console.error('Export consumption error:', error);
            throw error;
        }
    },

};

